# More SQL

Let's grab a fairly large dataset, load it into a database, and work with it.

## Getting your data

[Capital Bikeshare trip data](https://www.capitalbikeshare.com/trip-history-data) is a fun source of transactional data.  We can work with one quarter's data to show a few key concepts.

The following few cells should be feeling like old hat to you by now.

In [1]:
!wget https://www.capitalbikeshare.com/assets/files/trip-history-data/2013-Q1-Trips-History-Data.zip

--2015-10-27 16:39:56--  https://www.capitalbikeshare.com/assets/files/trip-history-data/2013-Q1-Trips-History-Data.zip
Resolving www.capitalbikeshare.com (www.capitalbikeshare.com)... 69.20.33.150
Connecting to www.capitalbikeshare.com (www.capitalbikeshare.com)|69.20.33.150|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6273788 (6.0M) [application/zip]
Saving to: ‘2013-Q1-Trips-History-Data.zip’

100%[======================================>] 6,273,788    491KB/s   in 11s    

2015-10-27 16:40:07 (577 KB/s) - ‘2013-Q1-Trips-History-Data.zip’ saved [6273788/6273788]



It's in a zip format, so unzip it:

In [2]:
!unzip 2013-Q1-Trips-History-Data.zip

Archive:  2013-Q1-Trips-History-Data.zip
  inflating: 2013-Q1-Trips-History-Data.csv  


How big is it?

In [3]:
!wc 2013-Q1-Trips-History-Data.csv

  396172  6116048 44445445 2013-Q1-Trips-History-Data.csv


What are its columns?

In [4]:
!csvcut -n 2013-Q1-Trips-History-Data.csv

  1: Duration
  2: Start date
  3: End date
  4: Start Station
  5: End Station
  6: Bike#
  7: Subscription Type


Okay, let's have a look.

In [5]:
!head -5 2013-Q1-Trips-History-Data.csv | csvlook

|-------------+-----------------+---------------+--------------------------------------+-------------------------------------------------------+--------+--------------------|
|  Duration   | Start date      | End date      | Start Station                        | End Station                                           | Bike#  | Subscription Type  |
|-------------+-----------------+---------------+--------------------------------------+-------------------------------------------------------+--------+--------------------|
|  0h 4m 58s  | 3/31/2013 23:59 | 4/1/2013 0:04 | Massachusetts Ave & Dupont Circle NW | New Hampshire Ave & T St NW [formerly 16th & U St NW] | W00347 | Casual             |
|  7h 45m 26s | 3/31/2013 23:58 | 4/1/2013 7:44 | Massachusetts Ave & Dupont Circle NW | New Hampshire Ave & T St NW [formerly 16th & U St NW] | W00550 | Casual             |
|  0h 3m 34s  | 3/31/2013 23:57 | 4/1/2013 0:01 | Crystal City Metro / 18th & Bell St  | 27th & Crystal Dr                   

Ah, that's kinda wordy.  Let's cut out that first column, which we can compute for ourselves later.

In [6]:
!head 2013-Q1-Trips-History-Data.csv | csvcut -C1 | csvlook

|------------------+-----------------+-----------------------------------------------+-------------------------------------------------------+--------+--------------------|
|  Start date      | End date        | Start Station                                 | End Station                                           | Bike#  | Subscription Type  |
|------------------+-----------------+-----------------------------------------------+-------------------------------------------------------+--------+--------------------|
|  3/31/2013 23:59 | 4/1/2013 0:04   | Massachusetts Ave & Dupont Circle NW          | New Hampshire Ave & T St NW [formerly 16th & U St NW] | W00347 | Casual             |
|  3/31/2013 23:58 | 4/1/2013 7:44   | Massachusetts Ave & Dupont Circle NW          | New Hampshire Ave & T St NW [formerly 16th & U St NW] | W00550 | Casual             |
|  3/31/2013 23:57 | 4/1/2013 0:01   | Crystal City Metro / 18th & Bell St           | 27th & Crystal Dr                               

That's a little bit cleaner, and the rest of the data should be useful.  Let's clean up the data by removing that column and renaming the headers so they're a little easier to query.

In [7]:
!csvcut -C1 2013-Q1-Trips-History-Data.csv | \
  header -r "start_date,end_date,start_station,end_station,bike_id,sub_type" \
  > bikeshare.csv

Make sure you haven't lost anything!

In [8]:
!wc bikeshare.csv

  396172  5323701 39974378 bikeshare.csv


## Prepping and loading data into the database

Alright, then, let's get loading.

In [9]:
%load_ext sql

/usr/local/lib/python2.7/dist-packages/IPython/config.py:13: ShimWarning: The `IPython.config` package has been deprecated. You should import from traitlets.config instead.
  "You should import from traitlets.config instead.", ShimWarning)
/usr/local/lib/python2.7/dist-packages/IPython/utils/traitlets.py:5: UserWarning: IPython.utils.traitlets has moved to a top-level traitlets package.
  warn("IPython.utils.traitlets has moved to a top-level traitlets package.")


**NOTE**: See a bunch of ShimWarnings with a pink background?  That's **normal**.  It's just a heads-up about ongoing changes to IPython/Jupyter code.  You can keep going.

First, we create a database in mysql.  **Note**: you can do the same thing on the command line by issuing the ```CREATE DATABASE``` command part before the pipe within the mysql shell, which you get to with the second part after the pipe.  Here we'll pipe the one into the other so it reads well in the notebook.

In [10]:
!echo "CREATE DATABASE bikedb" | mysql --user=mysqluser --password=mysqlpass 

Here's how we connect the notebook up to the mysql database using a username and password.  Remember that this shorthand version is possible thanks to the excellent [ipython-sql](https://github.com/catherinedevlin/ipython-sql) Jupyter extension that we're using, otherwise you'd have to establish the connection, get a cursor, etc., like you've done explicitly in python in your other class.  

Not that there's anything wrong with that.  

In [11]:
%sql mysql://mysqluser:mysqlpass@localhost/bikedb

u'Connected: mysqluser@bikedb'

Very easy, no?

First, clean up if we're not running this for the first time.

In [12]:
%%sql
DROP TABLE IF EXISTS bikeshare;

0 rows affected.


/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/default.py:450: Warning: Unknown table 'bikeshare'
  cursor.execute(statement, parameters)


[]

Next, create a table schema using DDL.

In [13]:
%%sql
CREATE TABLE bikeshare (
  start_date    DATETIME,
  end_date      DATETIME,
  start_station VARCHAR(100),
  end_station   VARCHAR(100),
  bike_id       CHAR(7),
  sub_type      CHAR(10)
  )

0 rows affected.


[]

Just to verify it worked:

In [14]:
%%sql
SELECT COUNT(*)
FROM bikeshare

1 rows affected.


[(0L,)]

It worked!  We just don't have any data in there yet.

Now we load the data using ```LOAD DATA INFILE```.  You can do pretty much the same thing from the bash shell using ```mysqlimport``` and a bunch of options.  It'll read better here in the notebook with the options spelled out.

Docs for ```LOAD DATA INFILE``` are available at <https://dev.mysql.com/doc/refman/5.1/en/load-data.html>.

**Note**: this assumes you've placed your bikeshare file in the directory ```/vagrant```.

**Note also**: I had to look up the [mysql date formatting docs](https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_date-format) to get this date format conversion correct.  It took me a few trials and errors before I got it right.  This is an **extremely common** thing to have to do if you ever spend time wrangling data - every system handles dates in its own way.

In [15]:
%%sql
LOAD DATA INFILE '/vagrant/bikeshare.csv'
REPLACE
INTO TABLE bikeshare
FIELDS TERMINATED BY ','
       OPTIONALLY ENCLOSED BY '"'
IGNORE 1 LINES
(@start_date, @end_date, start_station, end_station, bike_id, sub_type) 
SET start_date = STR_TO_DATE(@start_date, '%c/%e/%Y %k:%i'),
    end_date = STR_TO_DATE(@end_date, '%c/%e/%Y %k:%i')

396171 rows affected.


[]

**Note**: if the above command fails for you with a "file not found" error, please read [these notes about apparmor](http://ubuntuforums.org/showthread.php?t=822084&p=11253217#post11253217). Follow that advice, and add a line like it shows, e.g.:

    /vagrant/* r
    
...to the file, or whatever path you have your data on, reload apparmor, and try again.  I had to do this, and it worked perfectly after I made that change.

## Exploring your data

Now that we've loaded our data, or we think we have, let's just verify it.  Should be the same row count as what ```csvkit``` and ```wc``` gave us.

In [16]:
%%sql
SELECT COUNT(*) 
FROM bikeshare

1 rows affected.


[(396171L,)]

Looks good!  Let's look at the data a little.

In [17]:
%%sql
SELECT *
FROM bikeshare
LIMIT 5

5 rows affected.


[(datetime.datetime(2013, 3, 31, 23, 59), datetime.datetime(2013, 4, 1, 0, 4), 'Massachusetts Ave & Dupont Circle NW', 'New Hampshire Ave & T St NW [formerly 16th & U St NW]', 'W00347', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 58), datetime.datetime(2013, 4, 1, 7, 44), 'Massachusetts Ave & Dupont Circle NW', 'New Hampshire Ave & T St NW [formerly 16th & U St NW]', 'W00550', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 57), datetime.datetime(2013, 4, 1, 0, 1), 'Crystal City Metro / 18th & Bell St', '27th & Crystal Dr', 'W20193', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 4, 1, 0, 8), 'Georgia & New Hampshire Ave NW', '1st & Rhode Island Ave NW', 'W00209', 'Subscriber'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 3, 31, 23, 58), 'California St & Florida Ave NW', '21st & M St NW', 'W00646', 'Subscriber')]

How does MySQL construct this query, or more specifically, what's its execution plan?  We can find out with ```EXPLAIN```.

For more about how to read MySQL 5.5's query plan, see <https://dev.mysql.com/doc/refman/5.5/en/execution-plan-information.html>.

In [18]:
%%sql
EXPLAIN SELECT COUNT(*)
FROM bikeshare
LIMIT 5

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'ALL', None, None, None, None, 394019L, '')]

This says "using no keys, we're going to just scan roughly 395,390 rows, sans indexes, to answer this query."

In [19]:
%%sql
SELECT MAX(start_date)
FROM bikeshare

1 rows affected.


[(datetime.datetime(2013, 3, 31, 23, 59),)]

In [20]:
%%sql
EXPLAIN SELECT MAX(start_date)
FROM bikeshare

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'ALL', None, None, None, None, 394019L, '')]

Pretty much the same thing.  You can't get the max without looking at all of the values if there is no index.

In [21]:
%%sql
SELECT COUNT(*)
FROM bikeshare
WHERE start_station LIKE "%dupont%"

1 rows affected.


[(13498L,)]

In [22]:
%%sql
EXPLAIN SELECT COUNT(*)
FROM bikeshare
WHERE start_station LIKE "%dupont%"

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'ALL', None, None, None, None, 394019L, 'Using where')]

Now we see "using where" under "extra", so we know there's a filter operation, but that's about the only change.  What if we add more things to filter on?

In [23]:
%%sql
EXPLAIN SELECT start_station, end_station, COUNT(*)
FROM bikeshare
WHERE start_station LIKE "%dupont%"
AND end_station LIKE "%21st%"
AND start_date LIKE "2013-02-14%"
GROUP BY start_station, end_station
ORDER BY start_station, end_station

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'ALL', None, None, None, None, 394019L, 'Using where; Using temporary; Using filesort')]

Ah, some more info - it looks like it's using a temporary relation to store intermediate results, perhaps for the ```GROUP BY```, then a sort to handle ```ORDER BY```.

Still no indexes, though.  Let's change that.

In [24]:
%%sql
CREATE INDEX idx_start_station ON bikeshare (start_station)

0 rows affected.


[]

In [25]:
%%sql
EXPLAIN SELECT start_station, end_station, COUNT(*)
FROM bikeshare
WHERE start_station LIKE "21st%"
AND start_date LIKE "2013-02-14%"
GROUP BY start_station, end_station
ORDER BY start_station, end_station

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'range', 'idx_start_station', 'idx_start_station', '103', None, 21428L, 'Using where; Using temporary; Using filesort')]

I changed the query a little bit to use the index, do you see the difference?  It found search keys in the index, and the row count went down by an order of magnitude.  That's the power of indexes.

It helps even on simple queries like this.

In [26]:
%%sql
EXPLAIN SELECT DISTINCT start_station
FROM bikeshare
ORDER BY start_station

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'range', None, 'idx_start_station', '103', None, 201L, 'Using index for group-by')]

What's that 201 value for rows?  Maybe the actual count of distinct values.  We can test that:

In [27]:
%%sql
SELECT COUNT(*) 
FROM (
    SELECT DISTINCT start_station 
    FROM bikeshare
    ) made_up_subquery_alias_name

1 rows affected.


[(201L,)]

There you go, that's exactly the answer.

How about that ```MAX()``` query we tried a little while back?

In [28]:
%%sql
SELECT MAX(start_date)
FROM bikeshare

1 rows affected.


[(datetime.datetime(2013, 3, 31, 23, 59),)]

In [29]:
%%sql
EXPLAIN SELECT MAX(start_date)
FROM bikeshare

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'ALL', None, None, None, None, 394019L, '')]

Let's create another index on ```start_date``` to see what the effect on the query plan will be.

In [30]:
%%sql
CREATE INDEX idx_start_date ON bikeshare (start_date)

0 rows affected.


[]

In [31]:
%%sql
SELECT MAX(start_date)
FROM bikeshare

1 rows affected.


[(datetime.datetime(2013, 3, 31, 23, 59),)]

Same result, but...

In [32]:
%%sql
EXPLAIN SELECT MAX(start_date)
FROM bikeshare

1 rows affected.


[(1L, 'SIMPLE', None, None, None, None, None, None, None, 'Select tables optimized away')]

That's new!  In this case it doesn't have to look at any rows, it can just look at one end of the index.  We've optimized away the need to even look at the table.

Let's go back to ```COUNT()``` and try a few more things before we move on.

In [33]:
%%sql
EXPLAIN SELECT COUNT(*)
FROM bikeshare            

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'index', None, 'idx_start_date', '9', None, 394019L, 'Using index')]

In [34]:
%%sql
EXPLAIN SELECT COUNT(start_date)
FROM bikeshare   

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'index', None, 'idx_start_date', '9', None, 394019L, 'Using index')]

In [35]:
%%sql
EXPLAIN SELECT COUNT(end_date)
FROM bikeshare   

1 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'ALL', None, None, None, None, 394019L, '')]

Do you see what happened there?

## Normalizing attributes

Let's look at a few tasks you might need to perform if you were normalizing this dataset.  Remember that in normalization, we reduce redundancy with the goal of consistency.

What's redundant?  Well, the station names for one.

In [37]:
%%sql
SELECT COUNT(DISTINCT start_station)
FROM bikeshare

1 rows affected.


[(201L,)]

In [38]:
%%sql
SELECT COUNT(DISTINCT end_station)
FROM bikeshare

1 rows affected.


[(202L,)]

Hmm, they're different.  Let's put them together.

In [46]:
%%sql
SELECT COUNT(DISTINCT station) FROM
(
  SELECT start_station AS station FROM bikeshare
    UNION
  SELECT end_station AS station FROM bikeshare
) a

1 rows affected.


[(202L,)]

We'll create a table to hold the names of stations.  Each station name should be represented once, and we'll assign a primary key to each in the form of a unique integer.

In [51]:
%%sql
CREATE TABLE station (
  id SMALLINT NOT NULL AUTO_INCREMENT,
  name VARCHAR(100),
  PRIMARY KEY (id)
)

0 rows affected.


[]

In [52]:
%%sql
SELECT COUNT(*) 
FROM station

1 rows affected.


[(0L,)]

Looks good.  Now we can load the data with an INSERT that draws from our previous query.  We can **skip** specifying the ```id``` because MySQL will do that for us.

**Note**: every database handles this issue its own way.  This is a nice convenience in MySQL; other database backends require more work.

In [55]:
%%sql
INSERT INTO station (name) 
SELECT DISTINCT station AS name
FROM
(
  SELECT start_station AS station FROM bikeshare
    UNION
  SELECT end_station AS station FROM bikeshare
) a

202 rows affected.


[]

In [56]:
%%sql
SELECT * 
FROM station
LIMIT 10

10 rows affected.


[(1, '10th & E St NW'),
 (2, '10th & Monroe St NE'),
 (3, '10th & U St NW'),
 (4, '10th St & Constitution Ave NW'),
 (5, '11th & F St NW'),
 (6, '11th & H St NE'),
 (7, '11th & K St NW'),
 (8, '11th & Kenyon St NW'),
 (9, '12th & Army Navy Dr'),
 (10, '12th & L St NW')]

It worked.  Now we can update the bikeshare table to add columns for station identifiers.

In [57]:
%%sql
ALTER TABLE bikeshare
ADD COLUMN start_station_id SMALLINT
AFTER start_station

396171 rows affected.


[]

Looks good.  But what exactly just happened?

In [58]:
%%sql
DESCRIBE bikeshare

7 rows affected.


[('start_date', 'datetime', 'YES', 'MUL', None, ''),
 ('end_date', 'datetime', 'YES', '', None, ''),
 ('start_station', 'varchar(100)', 'YES', 'MUL', None, ''),
 ('start_station_id', 'smallint(6)', 'YES', '', None, ''),
 ('end_station', 'varchar(100)', 'YES', '', None, ''),
 ('bike_id', 'char(7)', 'YES', '', None, ''),
 ('sub_type', 'char(10)', 'YES', '', None, '')]

In [59]:
%%sql
SELECT * 
FROM bikeshare
LIMIT 5

5 rows affected.


[(datetime.datetime(2013, 3, 31, 23, 59), datetime.datetime(2013, 4, 1, 0, 4), 'Massachusetts Ave & Dupont Circle NW', None, 'New Hampshire Ave & T St NW [formerly 16th & U St NW]', 'W00347', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 58), datetime.datetime(2013, 4, 1, 7, 44), 'Massachusetts Ave & Dupont Circle NW', None, 'New Hampshire Ave & T St NW [formerly 16th & U St NW]', 'W00550', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 57), datetime.datetime(2013, 4, 1, 0, 1), 'Crystal City Metro / 18th & Bell St', None, '27th & Crystal Dr', 'W20193', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 4, 1, 0, 8), 'Georgia & New Hampshire Ave NW', None, '1st & Rhode Island Ave NW', 'W00209', 'Subscriber'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 3, 31, 23, 58), 'California St & Florida Ave NW', None, '21st & M St NW', 'W00646', 'Subscriber')]

What just happened?  Why are all the ```start_station_id``` values ```None```?

Let's fill in those values with our new identifiers from the ```station``` table.

In [65]:
%%sql
UPDATE bikeshare
INNER JOIN station
  ON bikeshare.start_station = station.name
SET bikeshare.start_station_id = station.id

396171 rows affected.


[]

In [66]:
%%sql
SELECT * FROM bikeshare LIMIT 5

5 rows affected.


[(datetime.datetime(2013, 3, 31, 23, 59), datetime.datetime(2013, 4, 1, 0, 4), 'Massachusetts Ave & Dupont Circle NW', 161, 'New Hampshire Ave & T St NW [formerly 16th & U St NW]', 'W00347', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 58), datetime.datetime(2013, 4, 1, 7, 44), 'Massachusetts Ave & Dupont Circle NW', 161, 'New Hampshire Ave & T St NW [formerly 16th & U St NW]', 'W00550', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 57), datetime.datetime(2013, 4, 1, 0, 1), 'Crystal City Metro / 18th & Bell St', 122, '27th & Crystal Dr', 'W20193', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 4, 1, 0, 8), 'Georgia & New Hampshire Ave NW', 135, '1st & Rhode Island Ave NW', 'W00209', 'Subscriber'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 3, 31, 23, 58), 'California St & Florida Ave NW', 107, '21st & M St NW', 'W00646', 'Subscriber')]

In [67]:
%%sql
SELECT * FROM station WHERE id = 161

1 rows affected.


[(161, 'Massachusetts Ave & Dupont Circle NW')]

Great, now we can drop ```start_station``` from ```bikeshare``` and save a lot of space.

In [68]:
%%sql
ALTER TABLE bikeshare
DROP COLUMN start_station

396171 rows affected.


[]

In [69]:
%%sql
DESCRIBE bikeshare

6 rows affected.


[('start_date', 'datetime', 'YES', 'MUL', None, ''),
 ('end_date', 'datetime', 'YES', '', None, ''),
 ('start_station_id', 'smallint(6)', 'YES', '', None, ''),
 ('end_station', 'varchar(100)', 'YES', '', None, ''),
 ('bike_id', 'char(7)', 'YES', '', None, ''),
 ('sub_type', 'char(10)', 'YES', '', None, '')]

In [70]:
%%sql
SELECT * FROM bikeshare LIMIT 5

5 rows affected.


[(datetime.datetime(2013, 3, 31, 23, 59), datetime.datetime(2013, 4, 1, 0, 4), 161, 'New Hampshire Ave & T St NW [formerly 16th & U St NW]', 'W00347', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 58), datetime.datetime(2013, 4, 1, 7, 44), 161, 'New Hampshire Ave & T St NW [formerly 16th & U St NW]', 'W00550', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 57), datetime.datetime(2013, 4, 1, 0, 1), 122, '27th & Crystal Dr', 'W20193', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 4, 1, 0, 8), 135, '1st & Rhode Island Ave NW', 'W00209', 'Subscriber'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 3, 31, 23, 58), 107, '21st & M St NW', 'W00646', 'Subscriber')]

Worked!

And we can repeat the process for ```end_station```.

In [71]:
%%sql
ALTER TABLE bikeshare
ADD COLUMN end_station_id SMALLINT
AFTER end_station

396171 rows affected.


[]

In [72]:
%%sql
UPDATE bikeshare
INNER JOIN station
  ON bikeshare.end_station = station.name
SET bikeshare.end_station_id = station.id

396171 rows affected.


[]

In [73]:
%%sql
ALTER TABLE bikeshare
DROP COLUMN end_station

396171 rows affected.


[]

In [74]:
%%sql
SELECT * FROM bikeshare LIMIT 5

5 rows affected.


[(datetime.datetime(2013, 3, 31, 23, 59), datetime.datetime(2013, 4, 1, 0, 4), 161, 172, 'W00347', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 58), datetime.datetime(2013, 4, 1, 7, 44), 161, 172, 'W00550', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 57), datetime.datetime(2013, 4, 1, 0, 1), 122, 67, 'W20193', 'Casual'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 4, 1, 0, 8), 135, 47, 'W00209', 'Subscriber'),
 (datetime.datetime(2013, 3, 31, 23, 52), datetime.datetime(2013, 3, 31, 23, 58), 107, 57, 'W00646', 'Subscriber')]

A lot leaner, right?


## JOINs and indexes

Now let's look at queries that return station names, thus requiring a ```JOIN``` across the two tables.  Keep in mind our two table schema.

In [77]:
%%sql
DESCRIBE station

2 rows affected.


[('id', 'smallint(6)', 'NO', 'PRI', None, 'auto_increment'),
 ('name', 'varchar(100)', 'YES', '', None, '')]

In [78]:
%%sql
DESCRIBE bikeshare

6 rows affected.


[('start_date', 'datetime', 'YES', 'MUL', None, ''),
 ('end_date', 'datetime', 'YES', '', None, ''),
 ('start_station_id', 'smallint(6)', 'YES', '', None, ''),
 ('end_station_id', 'smallint(6)', 'YES', '', None, ''),
 ('bike_id', 'char(7)', 'YES', '', None, ''),
 ('sub_type', 'char(10)', 'YES', '', None, '')]

Let's try a basic query that looks for the most busy station pairs.

In [87]:
%%sql
SELECT COUNT(*) AS c, start_station_id, end_station_id
FROM bikeshare
GROUP BY start_station_id, end_station_id
ORDER BY c DESC
LIMIT 5

5 rows affected.


[(1314L, 125, 154),
 (1257L, 154, 125),
 (979L, 94, 108),
 (908L, 172, 161),
 (895L, 108, 94)]

Now let's liven it up by joining to ```station``` and including station names.  We'll need to join twice, using two aliases.

Worked just fine.  Let's look under the hood, though.

In [97]:
%%sql
SELECT COUNT(*) AS c, station_1.name AS start_station, station_2.name AS end_station
FROM bikeshare, station AS station_1, station AS station_2
WHERE station_1.id = bikeshare.start_station_id
  AND station_2.id = bikeshare.end_station_id
GROUP BY bikeshare.start_station_id, bikeshare.end_station_id
ORDER BY c DESC
LIMIT 5

5 rows affected.


[(1314L, 'Eastern Market Metro / Pennsylvania Ave & 7th St SE', 'Lincoln Park / 13th & East Capitol St NE '),
 (1257L, 'Lincoln Park / 13th & East Capitol St NE ', 'Eastern Market Metro / Pennsylvania Ave & 7th St SE'),
 (979L, 'Adams Mill & Columbia Rd NW', 'Calvert St & Woodley Pl NW'),
 (908L, 'New Hampshire Ave & T St NW [formerly 16th & U St NW]', 'Massachusetts Ave & Dupont Circle NW'),
 (895L, 'Calvert St & Woodley Pl NW', 'Adams Mill & Columbia Rd NW')]

Looks good, and it's in my neighborhood. :)

Let's look at the query plan for all this:

In [101]:
%%sql
EXPLAIN SELECT COUNT(*) AS c, station_1.name AS start_station, station_2.name AS end_station
FROM station AS station_1, station AS station_2, bikeshare
WHERE bikeshare.start_station_id = station_1.id
  AND bikeshare.end_station_id = station_2.id
GROUP BY bikeshare.start_station_id, bikeshare.end_station_id
ORDER BY c DESC
LIMIT 5

3 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'ALL', None, None, None, None, 396342L, 'Using temporary; Using filesort'),
 (1L, 'SIMPLE', 'station_2', 'eq_ref', 'PRIMARY', 'PRIMARY', '2', 'bikedb.bikeshare.end_station_id', 1L, ''),
 (1L, 'SIMPLE', 'station_1', 'eq_ref', 'PRIMARY', 'PRIMARY', '2', 'bikedb.bikeshare.start_station_id', 1L, '')]

Not bad, but it's doing a full table scan on ```bikeshare```.  Let's see if some indexes would help with the two joins.

In [102]:
%%sql
CREATE INDEX idx_start_station_id ON bikeshare (start_station_id)

0 rows affected.


[]

In [103]:
%%sql
CREATE INDEX idx_end_station_id ON bikeshare (end_station_id)

0 rows affected.


[]

In [104]:
%%sql
EXPLAIN SELECT COUNT(*) AS c, station_1.name AS s1_name, station_2.name AS s2_name
FROM bikeshare, station AS station_1, station AS station_2
WHERE station_1.id = bikeshare.start_station_id
  AND station_2.id = bikeshare.end_station_id
GROUP BY bikeshare.start_station_id, bikeshare.end_station_id
ORDER BY c DESC
LIMIT 5

3 rows affected.


[(1L, 'SIMPLE', 'bikeshare', 'ALL', 'idx_start_station_id,idx_end_station_id', None, None, None, 396342L, 'Using temporary; Using filesort'),
 (1L, 'SIMPLE', 'station_2', 'eq_ref', 'PRIMARY', 'PRIMARY', '2', 'bikedb.bikeshare.end_station_id', 1L, ''),
 (1L, 'SIMPLE', 'station_1', 'eq_ref', 'PRIMARY', 'PRIMARY', '2', 'bikedb.bikeshare.start_station_id', 1L, '')]

Well, it's hard to say how much better this will perform without a lot more data. A ```COUNT``` operation simply needs to be able to count everything, if the level of granularity it's counting doesn't already have an easy lookup like we saw before.  Sometimes you just don't feel the pain of scale until you hit a scaling threshold that varies with the shape of your data.

But - see the ```possible_keys``` in the first row?  That means the optimizer sees the indexes present and will attempt to use those to at least organize the query a little better than it would be able to do without them.

Let's try one more thing - we can create an index on multiple columns that matches our query more precisely.  It's inefficient tot look up one column, then another, after all, we're looking for combinations of both.  A multiple column index can precompute that.

In [105]:
%%sql
CREATE INDEX idx_stations ON bikeshare (start_station_id, end_station_id)

0 rows affected.


[]

In [106]:
%%sql
EXPLAIN SELECT COUNT(*) AS c, station_1.name AS s1_name, station_2.name AS s2_name
FROM bikeshare, station AS station_1, station AS station_2
WHERE station_1.id = bikeshare.start_station_id
  AND station_2.id = bikeshare.end_station_id
GROUP BY bikeshare.start_station_id, bikeshare.end_station_id
ORDER BY c DESC
LIMIT 5

3 rows affected.


[(1L, 'SIMPLE', 'station_1', 'ALL', 'PRIMARY', None, None, None, 202L, 'Using temporary; Using filesort'),
 (1L, 'SIMPLE', 'bikeshare', 'ref', 'idx_start_station_id,idx_end_station_id,idx_stations', 'idx_stations', '3', 'bikedb.station_1.id', 1981L, 'Using where; Using index'),
 (1L, 'SIMPLE', 'station_2', 'eq_ref', 'PRIMARY', 'PRIMARY', '2', 'bikedb.bikeshare.end_station_id', 1L, '')]

Finally, looks like a big difference!